analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/ExplodedNode.cs

```diff
@@ -58,18 +58,37 @@ public int AddVisit()
     }
 
     public override int GetHashCode() =>
-        HashCode.Combine(programPointHash, State);
+        HashCode.Combine(programPointHash, State, FinallyPoint?.BlockIndex);
 
     public override bool Equals(object obj) =>
         Equals(obj as ExplodedNode);
 
     public bool Equals(ExplodedNode other) =>
         other is not null
         && other.programPointHash == programPointHash
-        && other.State.Equals(State);
+        && other.State.Equals(State)
+        && HasSameFinallyPointChain(other.FinallyPoint);
 
     public override string ToString() =>
         Operation.Instance is { } operation
             ? $"Block #{Block.Ordinal}, Operation #{index}, {operation.Serialize()}{Environment.NewLine}{State}"
             : $"Block #{Block.Ordinal}, Branching{Environment.NewLine}{State}";
+
+    private bool HasSameFinallyPointChain(FinallyPoint other)
+    {
+        var current = FinallyPoint;
+        while (current is not null && other is not null)
+        {
+            if (current.BlockIndex == other.BlockIndex && current.BranchDestination == other.BranchDestination)
```

Comment:
Suggestion: Could this line be an equals override in `FinallyPoint`?
And then you just check `current.Equals(other)`?
Looking at the codebase, this is the only point that you ever compare finally points.
Or more correctly maybe, the entire method should be inside `Equals`

Reply:
That was my initial implementation and I didn't go that way, because this logic is strictly bound to what ExplodedNode needs.

Such Equals would ignore that the branches are different, and are possibly from different source blocks. So I didn't want to use Equals for that.